// ngx-vis/esm5/ngx-vis.js
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { CUSTOM_ELEMENTS_SCHEMA, NgModule, NO_ERRORS_SCHEMA } from '@angular/core';
import { DataSet } from 'vis-network';
import { VisNetworkDirective } from './components/network/vis-network.directive';
import { VisNetworkService } from './components/network/vis-network.service';
import { VisTimelineDirective } from './components/timeline/vis-timeline.directive';
import { VisTimelineService } from './components/timeline/vis-timeline.service';
export { VisNetworkDirective, VisTimelineDirective, VisTimelineService, VisNetworkService, DataSet };
var VisModule = /** @class */ (function () {
    function VisModule() {
    }
    VisModule = tslib_1.__decorate([
        NgModule({
            declarations: [VisNetworkDirective, VisTimelineDirective],
            exports: [VisNetworkDirective, VisTimelineDirective],
            imports: [CommonModule],
            providers: [VisNetworkService, VisTimelineService],
            schemas: [CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA]
        })
    ], VisModule);
    return VisModule;
}());
export { VisModule };
